include/RAJA/policy/openmp/multi_reduce.hpp

```diff
@@ -0,0 +1,359 @@
+            m_data[index_data(bin, thread_idx, m_padded_bins, m_padded_threads)] = identity;
+          }
+        }
+      }
+    }
+  }
+
+  size_t num_bins() const { return m_num_bins; }
+
+  T identity() const { return m_identity; }
+
+  void combine(size_t bin, T const &val)
+  {
+    size_t thread_idx = omp_get_thread_num();
+    MultiReduceOp{}(m_data[index_data(bin, thread_idx, m_padded_bins, m_padded_threads)], val);
+  }
+
+  T get(size_t bin) const
+  {
+    ::RAJA::detail::HighAccuracyReduce<T, typename MultiReduceOp::operator_type>
+        reducer(m_identity);
+    for (size_t thread_idx = 0; thread_idx < m_max_threads; ++thread_idx) {
+      reducer.combine(m_data[index_data(bin, thread_idx, m_padded_bins, m_padded_threads)]);
+    }
+    return reducer.get_and_clear();
+  }
+
+private:
+  MultiReduceDataOMP const *m_parent;
+  size_t m_max_threads;
+  size_t m_num_bins;
+  size_t m_padded_threads;
+  size_t m_padded_bins;
+  T m_identity;
+  T* m_data;
+
+  static constexpr size_t pad_bins(size_t num_bins)
+  {
+    size_t num_cache_lines = RAJA_DIVIDE_CEILING_INT(num_bins*sizeof(T), RAJA::DATA_ALIGN);
+    return RAJA_DIVIDE_CEILING_INT(num_cache_lines * RAJA::DATA_ALIGN, sizeof(T));
+  }
+
+  static constexpr size_t pad_threads(size_t max_threads)
+  {
+    return max_threads;
+  }
+
+  static constexpr size_t index_data(size_t bin, size_t thread_idx,
+                                     size_t padded_bins, size_t RAJA_UNUSED_ARG(padded_threads))
+  {
+    return bin + thread_idx * padded_bins;
+  }
+
+  template < typename Container >
+  static T* create_data(Container const& container, T identity,
+                        size_t num_bins, size_t max_threads,
+                        size_t padded_bins, size_t padded_threads)
+  {
+    if (num_bins == size_t(0)) {
+      return nullptr;
+    }
+    auto data = RAJA::allocate_aligned_type<T>( RAJA::DATA_ALIGN, padded_threads*padded_bins*sizeof(T) );
+    if (max_threads > 0) {
+      {
+        size_t thread_idx = 0;
+        size_t bin = 0;
+        for (auto const& value : container) {
+          new(&data[index_data(bin, thread_idx, padded_bins, padded_threads)]) T(value);
+          ++bin;
+        }
+      }
+      for (size_t thread_idx = 1; thread_idx < max_threads; ++thread_idx) {
+        for (size_t bin = 0; bin < num_bins; ++bin) {
+          new(&data[index_data(bin, thread_idx, padded_bins, padded_threads)]) T(identity);
+        }
+      }
+    }
+    return data;
+  }
+
+  static void destroy_data(T*& data,
+                           size_t num_bins, size_t max_threads,
+                           size_t padded_bins, size_t padded_threads)
+  {
+    if (num_bins == size_t(0)) {
+      return;
+    }
+    for (size_t thread_idx = max_threads+1; thread_idx > 0; --thread_idx) {
+      for (size_t bin = num_bins; bin > 0; --bin) {
+        data[index_data(bin-1, thread_idx-1, padded_bins, padded_threads)].~T();
```

Comment:
Would it be more efficient to rewrite this so that you could use bin and thread_idx directly instead of repeatedly subtracting 1 in the loop? e.g. `for (size_t thread_idx = max_threads; thread_idx >= 0; --thread_idx)`

Reply:
The unsigned size_t used here would make that an infinite loop.

Reply:
I'm pretty sure this mimics what reverse iterators from the std library do so I'm not too concerned about performance. https://en.cppreference.com/w/cpp/iterator/reverse_iterator
Also for all the types that we would actually use here the destructor is trivial and this code is completely optimized out.